Bug fix: reliably detect UserMessage boundaries via typed spans

warpdev opened the pull request:
Fixes https://github.com/openai/codex/issues/2755
Replaces #2757 to fix root cause.

## Summary
  - Replace fragile string-based detection of user messages with typed metadata and span indices for backtracking.
  - Scope remains “User”-only typing; all other cells stay “Other”.
  - Backtracking logic is extracted and simplified with a consistent Esc/Enter flow across main and overlay views.

## Root cause
Backtrack relied on rendered text parsing (“user” header join/trim), which loses type info and misidentifies boundaries.

## Fix
- Add `MessageKind` (`User|Other`) and `MessageSpan` to `HistoryCell` (default keeps existing cells unchanged).
- Introduce `UserHistoryCell` and return it from `new_user_prompt` (transcript rendering unchanged).
- Replace `transcript_lines` with `transcript_cells` in App.
- Compute flattened transcript and absolute user spans on demand via `build_transcript_flattened()`. 

## Before

https://github.com/user-attachments/assets/64b262be-92c9-4159-b695-99e457cad56e

## After

https://github.com/user-attachments/assets/144b32a7-484e-462b-93ca-cfbb3cc66740

warpdev committed:
fix(tui): reliably detect UserMessage boundaries via typed spans

warpdev committed:
refact(tui): remove unused `_lines` params from backtrack helpers and update call sites/tests

warpdev committed:
test(tui): simplify backtrack helper tests to spans-only fixture

warpdev committed:
fix(tui): compute user spans from final inserted lines

warpdev commented:
Hi — This PR fixes a bug where `Resume` may select the wrong user message. When you have a moment, could you please take a look so others don’t run into it? Thank you! @nornagon-openai @easong-openai @bolinfest

nornagon-openai commented on codex-rs/tui/src/app.rs:
```diff
@@ -39,6 +39,8 @@ pub(crate) struct App {
     pub(crate) file_search: FileSearchManager,
 
     pub(crate) transcript_lines: Vec<Line<'static>>,
+    // Absolute [header..end) ranges for user messages within `transcript_lines`.
+    pub(crate) user_spans: Vec<(usize, usize)>,
```
can we try a version of this where instead of `transcript_lines` we store `transcript_cells`, and compute transcript_lines when we trigger transcript mode? then the transcript / backtrack view knows which cell is which and we can encapsulate this logic inside the backtrack view.

warpdev committed:
refact(tui): compute transcript on-demand; remove transcript_lines/user_spans and let overlay handle separation; update tests

warpdev committed:
refact(tui): dedupe Ctrl+T overlay and move transcript builder

warpdev commented:
Hi @nornagon-openai — could you please take another look when you have a moment?
I implemented the suggested `transcript_cells` approach and updated the PR description. Thanks!
